evennia/contrib/auditing/server.py

```diff
@@ -0,0 +1,266 @@
+    custom commands of your own that handle sensitive information.
+    
+    Installation:
+
+    Designate this class as the SERVER_SESSION_CLASS in `settings.py`, then set
+    some additional options concerning what to log and where to send it.
+    
+    settings.py:
+    SERVER_SESSION_CLASS = 'evennia.contrib.auditing.server.AuditedServerSession'
+    
+    # Where to send logs? Define the path to a module containing a function 
+    # called 'output()' you've written that accepts a dict object as its sole 
+    # argument. From that function you can store/forward the message received
+    # as you please. An example file-logger is below:
+    AUDIT_CALLBACK = 'evennia.contrib.auditing.examples'
+    
+    # Log all user input? Be ethical about this; it will log all private and 
+    # public communications between players and/or admins.
+    AUDIT_IN = True/False
+    
+    # Log all server output? This will result in logging of ALL system
+    # messages and ALL broadcasts to connected players, so on a busy MUD this 
+    # will be very voluminous!
+    AUDIT_OUT = True/False
+    
+    # What commands do you NOT want masked for sensitivity?
+    AUDIT_MASK_IGNORE = ['@ccreate', '@create']
+    
+    # What commands do you want to keep the first two terms of, masking the rest?
+    # This only triggers if there are more than two terms in the message.
+    AUDIT_MASK_KEEP_BIGRAM = ['create', 'connect', '@userpassword']
+    """
+    def audit(self, **kwargs):
+        """
+        Extracts messages and system data from a Session object upon message 
+        send or receive.
+    
+        Kwargs:
+            src (str): Source of data; 'client' or 'server'. Indicates direction.
+            text (list): Message sent from client to server.
+            text (str): Message from server back to client.
+    
+        Returns:
+            log (dict): Dictionary object containing parsed system and user data
+                related to this message.
+
+        """
+        # Get time at start of processing
+        time_obj = timezone.now()
+        time_str = str(time_obj)
+        
+        # Sanitize user input
+        session = self
+        src = kwargs.pop('src', '?')
+        bytes = 0
+        
+        if src == 'client':
+            try:
+                data = str(kwargs['text'][0][0])
+            except IndexError:
+                logger.log_err('Failed to parse client-submitted string!')
+                return False
+                
+        elif src == 'server':
+            # Server outputs can be unpredictable-- sometimes tuples, sometimes
+            # plain strings. Try to parse both.
+            try: 
+                if isinstance(kwargs.get('text', ''), (tuple,list)):
+                    data = kwargs['text'][0]
+                elif not 'text' in kwargs and len(kwargs.keys()) == 1:
+                    data = kwargs.keys()[0]
+                else:
+                    data = str(kwargs['text'])
```

Comment:
This will not be correct for different types of input/outputfuncs. It's perfectly possible for the system to receive (for example) `msg(text="Hello", prompt="hp=20|st=10|ma=15", pane=2)` which will translate to more keys in `**kwargs`. For OOB messages it's also possible to have no `text` key at all (since nothing should be displayed in the main window). In that case this line will lead to a traceback.

Reply:
For the sake of unpredictability, I think I will just cast kwargs to string and call it a day. Capture of server output was intended to be more of a diagnostic tool for admins anyway.

Reply:
Just casting raw kwargs will likely make things quite hard to read. It probably makes sense to identify the different input/outputfuncs when they are different to `text`. It's in fact a good candidate for a filter, to specify which input/outputfuncname to include/exclude in the logging. Things like regularly updating client-side OOB health-bar messages could quickly get very, very spammy.

Reply:
Argh...good point. You are not making this easy on me :)

Reply:
...and after trying things that way, it becomes much more difficult to reliably scrub commands for credentials.

My current compromise is to pop `text` from data when it's available, scrub it, and push it to the template as `text`. All other kwargs provided simply get passed through as `data` in the final object. I think this should accommodate everybody.